### PR TITLE
SRCH-1889 rate limit youtube

### PR DESCRIPTION
**Pull request by jmax-fearless**

### CODE REVIEW

**Comment by jmax-fearless** [comment]
The CodeClimate issues are existing ones that CC sees as new because I did some Rubocop cleanup on those lines. Requesting approval without fixing them.

File: spec/lib/tasks/youtube_spec.rb
Comment (MothOnMars) [inline spec/lib/tasks/youtube_spec.rb:8]: It took me a bit to figure out what was going on with this rejection line, but we shouldn't need that. This spec can be cleaned up quite a bit further while keeping Rubocop happy:
```ruby
describe 'YouTube rake tasks' do
  before(:all) { Rails.application.load_tasks }

  describe 'usasearch:youtube:refresh' do
    let(:task_name) { 'usasearch:youtube:refresh' }

    before { allow(YoutubeData).to receive(:refresh) }

    after { Rake::Task[task_name].reenable }

    it "has 'environment' as a prereq" do
      expect(Rake::Task[task_name].prerequisites).to include('environment')
    end

    it 'runs YoutubeData.refresh' do
      Rake::Task[task_name].invoke
      expect(YoutubeData).to have_received :refresh
    end
  end
end
```

File: spec/lib/importers/youtube_data_spec.rb
Comment (MothOnMars) [inline spec/lib/importers/youtube_data_spec.rb:14]: We can leave this for now, but one way to avoid this would be to set the `maximum_profile_updates_per_day` in the `test` environment to something extremely low, like 1 or 2. That way the loop would end quickly and automatically in specs, without the need to kill it. That could also simplify the specs, as we wouldn't have to stub that value to test the context `'when we have already updated the maximum number of profiles for today'`.

**Review by MothOnMars — CHANGES_REQUESTED** [review]
@jmax-fearless , sorry this review took forever. There are a lot of opportunities here to tighten up the code & specs. LMK if anything is unclear, and we can pair on the refactoring if necessary.

File: config/application.rb
Comment (jmax-fearless) [reply config/application.rb:102]: I went down that path initially, but switched to simply loading the YAML directly because this doesn't seem like a setting that needs to be different in development/test/production, and `config_for` requires that we supply settings per-environment.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:4]: The idea was to supply a default in case there is no config file setting.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:17]: Excellent point. Changed.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:24]: As discussed on Slack and summarized in the ticket, we're not changing the loop, so this code needs to stay.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:11]: Per our discussion on Slack, it's not that simple; the task is not set up as a cron job, and would need to be for this to work.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:38]: Nope. The .utc is the missing part. Otherwise I would've implemented it as a scope.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:43]: The idea is to provide a default in case the config setting is missing.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:52]: Moved the logging out and combined the boolean operations.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:10]: I'm fine with disabling that cop, and will revert the code back to `recieve_message_chain`

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:46]: Removed.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:42]: I believe the code as written is significantly clearer. In particular, I don't like embedding an arbitrary number which may well change.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:90]: I'll remove the specs you don't like. As discussed above, I see no way to do this with a scope; the issue is that the check needs to be done in UTC.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:317]: Removed.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:14]: An excellent idea, once we actually break the infinite loop, but as agreed, that's out of scope for this ticket.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb]: Non-issue; I removed the offending code while rolling back the refactorings for `receive_method_chain`

File: spec/lib/tasks/youtube_spec.rb
Comment (jmax-fearless) [reply spec/lib/tasks/youtube_spec.rb:8]: Changed.

Comment (jmax-fearless) [reply spec/lib/tasks/youtube_spec.rb]: Backed out the change; doing this breaks our code coverage tool, and it incorrectly reports uncovered code in unrelated rake tasks. Precisely which ones depends on RSpec's ordering; any task's specs which ran before this one will show as not being covered by tests. It looks like the call to load_tasks resets the coverage counters for the tasks.

Comment (MothOnMars) [reply spec/lib/tasks/youtube_spec.rb:8]: In that case, let's simplify things and back out the changes to this file entirely, as no changes are necessary for this story.

File: config/application.rb
Comment (MothOnMars) [reply config/application.rb:102]: My thinking for using a YAML file was that we might use a different setting in test/dev environments, and also that it would lay the groundwork for eventually moving the rest of the Youtube settings out of `config/secrets.yml` and into `config/youtube.yml`. However, we haven't identified a need to do either of those things at this point, so I think we should keep it simple and just use a constant and no config file at all.

File: lib/importers/youtube_data.rb
Comment (MothOnMars) [reply lib/importers/youtube_data.rb:4]: See note above. Let's just use a constant and nix the config file, which is not needed at this point in time:
```
MAXIMUM_PROFILE_UPDATES_PER_DAY = 300
```

File: lib/importers/youtube_data.rb
Comment (MothOnMars) [reply lib/importers/youtube_data.rb:24]: 👍

File: lib/importers/youtube_data.rb
Comment (MothOnMars) [reply lib/importers/youtube_data.rb:11]: 👍 
For future reference: https://cm-jira.usa.gov/browse/SRCH-1983

File: lib/importers/youtube_data.rb
Comment (MothOnMars) [reply lib/importers/youtube_data.rb:38]: Production time is UTC, so `Date.today` would work. However, I see value in being explicit about `utc`. You can still use that in a scope, as long as it's in a lambda:
```ruby
app/models/youtube_profile.rb

scope :updated_today, -> { where('updated_at > ?', Time.now.utc.to_date) }
```
Or without a scope:
```ruby
YoutubeProfile.where('updated_at > ?', Time.now.utc.to_date).count
```

File: lib/importers/youtube_data.rb
Comment (MothOnMars) [reply lib/importers/youtube_data.rb:43]: As noted above, let's simplify this with a single constant.

File: spec/lib/importers/youtube_data_spec.rb
Comment (MothOnMars) [reply spec/lib/importers/youtube_data_spec.rb]: OK - go ahead and add a subtask to this story to disable that cop in searchgov_style. We can just approve any failures for that cop before that change is merged.

File: spec/lib/importers/youtube_data_spec.rb
Comment (MothOnMars) [reply spec/lib/importers/youtube_data_spec.rb]: see note above on how to do this with a scope

**Comment by MothOnMars** [comment]
@jmax-fearless , I responded to the comments so far. There are also some file conflicts that will need to be resolved.

File: config/application.rb
Comment (jmax-fearless) [reply config/application.rb:102]: OK.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:4]: OK

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:38]: Cool. Enthusiastically accepted; I'll switch it over to a scope.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:43]: OK

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb]: OK.

File: spec/lib/tasks/youtube_spec.rb
Comment (jmax-fearless) [reply spec/lib/tasks/youtube_spec.rb:8]: OK

Comment (jmax-fearless) [reply spec/lib/tasks/youtube_spec.rb:8]: I left the # frozen_string_literal: true in, but backed out all of the other changes.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb]: Done.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb]: This code has been simplified somewhat by the switch to using the MAXIMUM_PROFILE_UPDATES_PER_DAY constant, instead of the config file; are you OK with it now?

Comment (MothOnMars) [reply spec/lib/importers/youtube_data_spec.rb]: My preference is still to be explicit, specifically _because_ it's "arbitrary number which may well change". Being explicit in the spec is what protects us from unintentional changes, such as someone accidentally changing `300` to `3000` in the code itself. That said, I leave it up to you.

File: lib/importers/youtube_data.rb
Comment (jmax-fearless) [reply lib/importers/youtube_data.rb:33]: Left there; I feel that the method name is moderately clearer.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:114]: Changed.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb]: Changed.

File: app/models/youtube_profile.rb
Comment (jmax-fearless) [reply app/models/youtube_profile.rb:20]: Added.

File: spec/lib/importers/youtube_data_spec.rb
Comment (jmax-fearless) [reply spec/lib/importers/youtube_data_spec.rb:60]: Keeping, since I'm keeping the method. I would agree with you 100% if I were killing it.

**Comment by jmax-fearless** [comment]
@MothOnMars - I kept the method `number_of_profiles_imported_today`, and therefore kept distinct specs for the method and the scope it relies on. Made the other changes you requested.

File: spec/models/youtube_profile_spec.rb
Comment (MothOnMars) [inline spec/models/youtube_profile_spec.rb:45]: should be:
`context 'when no profiles have been imported' do` (not `updated`)

File: spec/models/youtube_profile_spec.rb
Comment (MothOnMars) [inline spec/models/youtube_profile_spec.rb:57]: This spec should be updated to test the scope more explicitly. (The context should also be updated to specify `imported` rather than `updated`.) As currently written, the scope could include all profiles (the fact that the other fixtures in the DB don't include an `imported_at` value is not obvious in the spec). We should clarify that profiles with older `imported_at` values aren't included, i.e.:
```ruby
  describe 'imported_today' do
    context 'when no profiles have been imported' do
      it 'returns no profiles' do
        expect(described_class.imported_today).to be_empty
      end
    end

    context 'when profiles have been imported' do
      before do
        described_class.first.update!(imported_at: Time.now)
        described_class.last.update!(imported_at: 1.day.ago)
      end

      it 'includes profiles that were imported today' do
        expect(described_class.imported_today).to eq([described_class.first])
      end
    end
  end
```

**Review by MothOnMars — CHANGES_REQUESTED** [review]
Just a couple of tweaks to the new spec

File: spec/models/youtube_profile_spec.rb
Comment (jmax-fearless) [reply spec/models/youtube_profile_spec.rb:57]: Changed.

**Review by MothOnMars — APPROVED** [review]
I just realized that another one of the CodeClimate failures was for new code, namely the use of `Time.now` in:
```ruby
before { YoutubeProfile.first.update!(imported_at: Time.now) }
```
I'm approving so we can wrap this up, but let's be careful not to assume that CC failures are pre-existing.

**Pull request merged**
